### PR TITLE
[C++ API Parity] [Optimizers] added closure to optimizers

### PR DESCRIPTION
**Pull request by anjali411**
This PR is BC breaking in the following way:
`step` function in `Optimizer` class now takes an optional `LossClosure` closure function and returns a `Tensor`. If you had a custom optimizer class defined as:
```cpp
struct MyOptimizer : Optimizer {
  using Optimizer::Optimizer;
  void step() override {...}
};
```
you would need to change the class definition to conform to the new signature of step function and update your class definition as follows:

```cpp
struct MyOptimizer : Optimizer {
  using Optimizer::Optimizer;
  torch::Tensor step(LossClosure closure = nullptr) override { ... }
};
```

### CODE REVIEW

File: test/cpp/api/serialize.cpp [hunk 1]
```diff
@@ -100,7 +100,8 @@ void test_serialize_optimizer(DerivedOptimizerOptions options) {
     optimizer.zero_grad();
     auto y = model->forward(x).sum();
     y.backward();
-    optimizer.step();
+    auto closure = []() { return torch::tensor({10}); };
+    optimizer.step(closure);
```
Comment (yf225) [inline test/cpp/api/serialize.cpp]: Wondering would it work if we do `optimizer.step(nullptr)` instead? It seems to be simpler than having to create a closure that we don't actually need

File: torch/csrc/api/include/torch/optim/optimizer.h [hunk 2]
```diff
@@ -199,8 +199,9 @@ TORCH_API serialize::InputArchive& operator>>(
 /// according to the concrete optimization algorithm.
 class Optimizer : public detail::OptimizerBase {
  public:
-  using detail::OptimizerBase::OptimizerBase;
-  virtual void step() = 0;
+   using LossClosure = std::function<Tensor()>;
```
Comment (yf225) [inline torch/csrc/api/include/torch/optim/optimizer.h]: maybe we can add the comment above: `/// A loss function closure, which is expected to return the loss value.`

**Review by yf225 — APPROVED** [review]
Thanks a lot for the great work! As a follow-up PR, it would be awesome to add simple tests to check that passing closure to step function is working for all optimizers changed in this PR :)

**Comment by yf225** [comment]
Also this PR is BC-breaking - it would be great to write in the PR description why it is BC-breaking ;)
